### ISSUE TITLE
Turbo is very slow to start (80 seconds)

### ISSUE THREAD

**Issue by renchap** [vercel/turborepo#667]
### What version of Turborepo are you using?

1.1.2

### What package manager are you using / does the bug impact?

Yarn v2/v3 (node_modules linker only)

### What operating system are you using?

Mac

### Describe the Bug

I am trying to switch our monorepo task manager from Lerna to Turbo.
There are 3 very simple pipelines, without any dependencies or outputs (`check-typescript`, `lint` and `test`)

`turbo run` is very slow, even for tasks only ran in 2 packages and with all results cached:

```
yarn run turbo run test --continue
• Packages in scope: …
• Running test in 7 packages
…

 Tasks:    2 successful, 2 total
Cached:    2 cached, 2 total
  Time:    1m21.491s >>> FULL TURBO
```

Running the tasks in each repo is quick. The waiting time occurs before Turbo outputs anything.

### Expected Behavior

Turbo does not take 1.2 minutes

### To Reproduce

I can reproduce it consistently locally, but not sure what is triggering it. Is it a way to know what is Turbo doing before launching the tasks?

I can privately provide traces & cpu graphs, but from looking at it quickly I think it comes from hashing the files. The repo contains simple JS packages, but also a Rails app and a React Native app and the monorepo directory is 12 GB and 288071 files in my workstation (contains local development data, Android & iOS build files…).

I did not find a way to tell Turbo to ignore some subdirectories.

**Comment by renchap** [comment]
Here is a flamegraph of a run, most of the time is spent in reading files

<img width="2403" alt="image" src="https://user-images.githubusercontent.com/42070/152219601-9393283c-8851-4018-b43c-85252f06dbd1.png">

**Comment by jaredpalmer** [comment]
Check that your `workspaces` definition is as precise as possible. `turbo` is likely walking directories needlessly.

**Comment by renchap** [comment]
Here are my current workspaces:
```
  "workspaces": {
    "packages": [
      "packages/*"
    ]
  },
```

This covers all 7 JS packages in the mono-repo, so I cant be more precise.

If I switch to only one package

```
  "workspaces": {
    "packages": [
      "packages/core",
    ]
  },
```

Turbo is much faster.

If I add the React Native app:

```
  "workspaces": {
    "packages": {
       "packages/core",
      "packages/mobile"
    ]
  },
```

Turbo takes 45 seconds.

If I add the Rails (+ webpack/react) app, then any `turbo run` command takes 80+ seconds.

Both the Rails & React Native app contain a `package.json` and have tasks I want to run using Turbo.

Would it be possible to have Turbo ignore some sub-directories in the hashing algorithm, using either a configuration flag or a `.turboignore`?

**Comment by elado** [comment]
Same issue here. Repo is 30GB, of which `.git` is ~22GB.

On an Intel machine it would take ~1m, on M1 Max ~25s. This is unexpected.

![image](https://user-images.githubusercontent.com/51488/157173620-5b801379-c02a-4e4a-a1e8-db5339cd5903.png)

**Comment by elado** [comment]
Going through the debugger, this line takes a long time to run, when there's a presence of many (>500k) files in the repo.

https://github.com/vercel/turborepo/blob/main/cli/internal/globby/globby.go#L27

```
_ = fs.Walk(basePath, func(p string, isDir bool) error {
```


The rest (on a `FULL TURBO` run) is instantaneous.

**Comment by jaredpalmer** [comment]
Will have team investigate

**Comment by crubier** [comment]
Same here, relatively large repo, takes ~40s to start.

The tool yarn.build ( https://github.com/ojkelly/yarn.build ) is much faster in my case because it is based on modification date instead of file hashes.

**Comment by nathanhammond** [comment]
Hey y'all, with #1108 landing we should have made some improvements to this situation by only walking the files that are possibly included instead of the entire root path.

If you have a moment we would greatly appreciate hearing if this addresses the issues you've encountered. (We still have numerous further optimizations available to us but we're hoping to get file system performance into a stable state for further triage.)

**Comment by crubier** [comment]
In my case #1108 does not seem to change much.

I tested both:

  - Before #1108 (v1.2.4 from npm)
  - After #1108 (version I built locally from [this commit](https://github.com/vercel/turborepo/commit/db68365b0bb67ef79dad5000ab573e5a4a5d3e49))

Here is my report:

- Build one package with `FULL TURBO`: 
  - Before: avg `14s` (runs: `14.398s`, `13.91s`, `18.135s`, `14.025s`)
  - After: avg `14s` (runs: `14.289s`, `14.296s`, `14.311s`, `14.282s`)
- Build all packages with `FULL TURBO`: 
  - Before: avg `30s` (runs:  `27.924s`,  `16.303s`, `45.543s`, `37.422s`)
  - After: avg `30s`  (runs: `34.889s` , `22.946s`, `30.045s`, `28.854s`)

**Comment by elado** [comment]
Woohoo! 1.2.6 fixes this issue for me. On the same project the 40s run is now 2-5s. Thank you! 🙏

**Comment by nathanhammond** [comment]
@renchap Thank you for all of the information you provided, with all of your information and @elado's response I feel pretty confident that we have likely addressed the issue you were encountering. If you take another look with turbo@1.2.6 and are still encountering issues, please let me know and I'll reopen.

@crubier Can you file a new issue with more details about your setup? I'd like to try and track down what is going on with your problem. I suspect that it's something else and I will need to investigate the root cause.

**Comment by nathanhammond** [comment]
@crubier If you can also pass me the flame chart that this script generates I'll be able to quickly identify the problem:

```
turbo run build --cpuprofile=cpu.prof
go tool pprof -http=:8080 cpu.prof
```

**Comment by ObliviousHarmony** [comment]
I ran into this problem too @nathanhammond and it was a consequence of [the heavy-handed glob patterns suggested by PNPM's documentation](https://pnpm.io/pnpm-workspace_yaml).

<img width="1759" alt="Screen%20Shot%20on%202022-05-16%20at%2012-09-25" src="https://user-images.githubusercontent.com/6451942/168668269-05037902-9056-4662-9c0c-7046ed23d39e.png">

Once I changed the glob pattern to `packages/*` performance improved and execution became instantaneous.
